src/v/storage/segment_reader.cc

```diff
@@ -89,4 +208,47 @@ std::ostream& operator<<(std::ostream& os, const segment_reader_ptr& seg) {
     }
     return os << "{{log_segment: null}}";
 }
+
+segment_reader_handle::segment_reader_handle(segment_reader* parent)
+  : _parent(parent) {
+    _parent->_streams.push_back(*this);
+}
+
+segment_reader_handle::~segment_reader_handle() {
+    vassert(!_stream.has_value(), "Must close before destroying");
+    vassert(_parent == nullptr, "Must close before destroying");
+}
+
+ss::future<> segment_reader_handle::close() {
+    auto f = ss::now();
+    if (_stream) {
+        co_await _stream.value().close();
+        _stream = std::nullopt;
+    }
+    _hook.unlink();
+
+    if (_parent) {
+        co_await _parent->put();
+        _parent = nullptr;
+    }
+}
+
+void segment_reader_handle::operator=(segment_reader_handle&& rhs) noexcept {
+    if (_parent) {
+        // Where we move-assign a handle into a handle that's already,
+        // it's to reset a stream on the same underlying segment_reader,
+        // so we can be certain that the put() will not reduce the
+        // file handle refcount to zero, as `rhs` holds a reference.
+        vlog(
+          stlog.trace,
+          "Backgrounding put to {} {}",
+          _parent->_filename,
+          fmt::ptr(_parent));
+        ssx::background = _parent->put();
+    }
+    _stream = std::exchange(rhs._stream, std::nullopt);
+    _parent = std::exchange(rhs._parent, nullptr);
```

Comment:
do move-assignment operators benefit from guards against self move like `x = std::move(x)`?

Reply:
In cases like this where we would never expect to self-assign: yes, I think so.  Added an assertion.